### DISCUSSION TITLE
Integration of Rust based libraries into Bisq 2

### DISCUSSION THREAD

**Discussion by HenrikJannsen** [bisq-network/bisq2#2860] [General]
For the MuSig protocol [1] we will likely use the Rust based BDK library and likely some native C libraries for MuSig.

To avoid security risks for Bisq as well as to avoid new dependencies and complexity for the build system, we could run those as native applications and communicate via GRPC or TCP sockets [2] with the java desktop app. By that we can clearly limit the risk exposure to the network messages defined.

To have an independent Rust repository for that which integrates BDK, the C-native libs and the low level API for what we need, this project might become also attractive for other developers. This might be also a good option for attracting Rust developers to join. 

[1] https://github.com/bisq-network/proposals/issues/456
[2] Similar model is used for the webcam app

**Comment by cbeams** [comment]
In such a discussion people will naturally want to know whether the BDK ships its own JVM (JNI) library bindings. They do, and they're here: https://github.com/bitcoindevkit/bdk-ffi/tree/master/bdk-jvm.

**Reply by cbeams** [reply]
Correction: I wrote JNI above but should have written _JNA_, as that is actually what BDK's bindings are based on.

- https://github.com/java-native-access/jna
- https://en.wikipedia.org/wiki/Java_Native_Access

**Reply by HenrikJannsen** [reply]
Thanks for adding that. The primary motivation was not difficulty how to integrate it but rather to sandbox security risks and also to create more independence from the java project.

Beside that I think it might open up a new avenue into Rust for Bisq. 

Maybe in future once there are sufficient Rust devs around we could move other low level stuff like the P2P network also to Rust and by being a more independent library its more attractive to others to re-use and join efforts.
